pennylane/devices/default_qubit.py

```diff
@@ -204,6 +204,16 @@ def define_wire_map(self, wires):
         wire_map = zip(wires, consecutive_wires)
         return dict(wire_map)
 
+    # pylint: disable=arguments-differ
+    def _get_batch_size(self, tensor, expected_shape, expected_size):
+        """Determine whether a tensor has an additional batch dimension for broadcasting,
+        compared to an expected_shape."""
+        size = self._size(tensor)
+        if self._ndim(tensor) > len(expected_shape) or size > expected_size:
+            return size // expected_size
+
+        return None
+
     # pylint: disable=arguments-differ
     def apply(self, operations, rotations=None, **kwargs):
```

Comment:
Hi @dwierichs,

I am worried about lots of changes in the `apply` function as `lightning.qubit` (a child of `default.qubit`) overrides this function. Isn't it possible to apply broadcast at the QNode level (remaining some requirements for devices)? 

In general, as the current structure of device classes is not well-designed, I am also concerned that the changes in `default.qubit` beaks other custom plugins.

Reply:
Hey @chaeyeunpark 
Thanks for taking a look!
I am not entirely sure I understand the concern quite yet. Note that `lightning.qubit` will have to set the `supports_broadcasting` flag in the capabilities directory to `False` because it otherwise falsely inherits the `supports_broadcasting=True` flag from `DefaultQubit`. Once the internal methods like `apply` of lightning qubit also support broadcasting (if we go there, benchmarks are planned), the flag can be set to `True`.

Reply:
@chaeyeunpark is there a reason that `lightning.qubit` inherits from `default.qubit`? Going forward, it probably makes more sense to have it inherit from `QubitDevice`, to reduce chances of PL changes breaking lightning

Reply:
@josh146 That is one thing, but doing that also implies that this functionality is device-dependent and we need to implement it from the beginning. As a feature, I would like to see broadcasting can be supported by adding a single method in a device instead of changing all `apply` related functions. Thus it would be nice if we support it at a `QNode` level. One possibility is making a lightweight data structure modeling circuit+parameters (other than `tape` as the main reason why we need this PR is that `tape` is heavy) and adding a function in a device that executes this data structure (e.g. `apply_broadcast`). If `QNode` get parameters whose shape is `(N, n_params),` one may just construct this data structure and pass it to a device for broadcasting.

Reply:
I'm not sure I entirely agree. Even if tapes were super-lightweight and would come with a broadcastable data structure, executing a broadcasted state vector simulation will still be faster than a python for loop over the parameters, right?
I do not quite see how we would make use of off-loading the broadcasted dimension to numpy's C backend without allowing the apply methods to pass through broadcasted objects :thinking:

Reply:
The data structure should not bind the gates and parameters. It can just have a circuit structure and all parameters. For example, one may have a list of gate classes (not instances) and parameters for those gates, e.g. `{'ops':[qml.RX, qml,RY, qml,RZ], 'params'[([0.1.0.2,0.3,0.4], wires=[1,2]), ([0.1.0.2,0.3,0.4], wires=[3,4])]}`.